doc/man/man1/manochrome.1
.\" Automatically generated by Pod::Man 4.14 (Pod::Simple 3.42)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{\
.    if \nF \{\
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{\
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "manochrome 1"
.TH manochrome 1 "2024-08-18" "v0.1" "MANochrome"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
MANochrome
\&\- CSS and Javascript for HTML generated by pod2html(1).
.PP
manopod2html
\&\- Converts .pod files to .html files with MANochrome included.
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBmanopod2html\fR \fB\-t\fR theme [\fB\-i\fR input_file] \fB\-o\fR output_file [\fB\-f\fR]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
The goal is to present information in a man page like style with some
appeal to the eye. The \fIdark\fR theme comes closest to that goal.
The \fImonk\fR theme deviates a bit from the initially monochrome style while
still being simple and giving a pleasant reading experience. It's inspired
by <https://perlmonks.org>.
.PP
While the intention is mainly to convert \fI.pod\fR files, it is also possible to
convert man pages. For further information, please see the \*(L"\s-1EXAMPLES\*(R"\s0
section.
.PP
Nicely formatted \fI.pod\fR files work best and look great.
.SS "MANochrome"
.IX Subsection "MANochrome"
A set of cascading stylesheets (themes) and some Javascript which is
explained in the following sections.
.PP
\fIThemes\fR
.IX Subsection "Themes"
.PP
can be included in the <head> of an \s-1HTML\s0 file and will change the
look of the \s-1HTML\s0 page. 
They can be used in any \s-1HTML\s0 file but are intended to be used with \s-1HTML\s0
generated by \fBpod2html\fR\|(1).
.PP
Available themes are
.PP
.Vb 3
\&  light
\&  dark
\&  monk
.Ve
.PP
\fIJavascript\fR
.IX Subsection "Javascript"
.PP
is optional but must be inserted before the closing \f(CW\*(C`body\*(C'\fR tag.
.PP
\fIclass MANochrome\fR
.IX Subsection "class MANochrome"
.PP
The parent class, it is not initialized.
.PP
\fIclass MANochrome_pod\fR
.IX Subsection "class MANochrome_pod"
.PP
creates a sidebar with a fixed index where a link to the corresponding
section is \fBhighlighted\fR when \fBscrolling\fR through the page or when
\&\fBclicked\fR.
An Index must be present in the \s-1HTML\s0 for this to work. \fBpod2html\fR\|(1)
generates an index by default. Said index is an unordered list (\f(CW\*(C`<ul\*(C'\fR>) with
the \f(CW\*(C`id\*(C'\fR \f(CW"index"\fR.
.PP
The class must be initialized with some
Javascript after \fIjs/manochrome.js\fR was included.
.PP
.Vb 1
\&  const manochrome = new MANochrome_pod();
.Ve
.PP
\fIoptions\fR
.IX Subsection "options"
.PP
When initializing with \f(CW\*(C`new MANochrome_pod()\*(C'\fR an \fIoptions\fR object
can can be passed. This must be adjusted manually but is very easy. 
Take a look at \fIjs/init.js\fR, \*(L"manopod2html\*(R" inserts the content of this
file inside a script tag below the closing body tag.
.IP "\fBlogo_path\fR <string>" 4
.IX Item "logo_path <string>"
A path or \s-1URL\s0 to the image.
.IP "\fBlinks\fR <array>" 4
.IX Item "links <array>"
An array of objects like:
.Sp
.Vb 10
\&    "links": [
\&      {
\&        "url": "https://github.com/aaamendes/manochrome.git",
\&        "html": "<img src=\e"img/github\-mark\-white.svg\e""> Repo"
\&      },
\&      {
\&        "url": "https://perl.org",
\&        "html": "perl.org"
\&      },
\&      {
\&        "url": "https://perlmonks.org",
\&        "html": "perlmonks.org"
\&      }
\&    ]
.Ve
.PP
\&\*(L"manopod2html\*(R" can do all of this for you.
.SS "manopod2html"
.IX Subsection "manopod2html"
This perl script uses \f(CW\*(C`Pod::Html\*(C'\fR to generate \s-1HTML\s0 from a \fI.pod\fR file.
If used without the \fB\-f\fR flag, it will create a directory
\&\fImanochrome_build\fR with the structure
.PP
.Vb 6
\&        manochrome_build
\&        +\-\- css
\&        |\ \  \e\-\- manochrome_dark.css
\&        +\-\- index.html
\&        \e\-\- js
\&            \e\-\- manochrome.js
.Ve
.PP
where <theme> is the theme chosen with the \fB\-t\fR flag and
<output.html> is the output file name chosen with \fB\-o\fR.
.PP
\fI\s-1OPTIONS\s0\fR
.IX Subsection "OPTIONS"
.IP "\fB\-t\fR <theme>" 4
.IX Item "-t <theme>"
Pass the name of a theme.
.IP "\fB\-i\fR <input_file>" 4
.IX Item "-i <input_file>"
Pass the input file.
.Sp
If this flag is ommitted, read from \s-1STDIN.\s0
.IP "\fB\-o\fR <output_file>" 4
.IX Item "-o <output_file>"
Pass the output file.
.IP "\fB\-f\fR" 4
.IX Item "-f"
Build a single file with css and javascript included.
.SH "EXAMPLES"
.IX Header "EXAMPLES"
.SS "Build from \s-1POD\s0"
.IX Subsection "Build from POD"
Build \f(CW\*(C`index.html\*(C'\fR from \f(CW\*(C`Readme.pod\*(C'\fR in the repository
<https://github.com/aaamendes/manochrome.git>
.PP
.Vb 1
\&  manopod2html \-i Readme.pod \-o index.html \-t dark \-f
.Ve
.SS "Build from \s-1MAN\s0"
.IX Subsection "Build from MAN"
The tool \fBrman\fR\|(1) can create \fI.pod\fR files from man pages.
We use it in combination with manopod2html:
.PP
.Vb 2
\&  man man | rman \-f POD | \e
\&  manopod2html \-o man.html \-t dark \-f
.Ve
.SS "manochrome options"
.IX Subsection "manochrome options"
Add this script to the bottom of your html before the closing body tag to
initialize \f(CW\*(C`MANochrome_pod\*(C'\fR with options.
.PP
.Vb 3
\&  <script>
\&    () => {
\&      \*(Aquse strict\*(Aq;
\&    
\&      const options = {
\&        /*
\&         * Add a logo to top of the sidebar.
\&         */
\&        "logo_path": "img/logo.png",
\&        /*
\&         * Add links to the "LINKS" section at the bottom of the sidebar.
\&         * The "LINKS" section will only be created if this option is passed.
\&         */
\&        "links": [
\&          {
\&            "url": "https://github.com/aaamendes/manochrome.git",
\&            "html": "<img src=\e"img/github\-mark\-white.svg\e"> Repo"
\&          },
\&          {
\&            "url": "https://perl.org",
\&            "html": "perl.org"
\&          },
\&          {
\&            "url": "https://perlmonks.org",
\&            "html": "perlmonks.org"
\&          }
\&        ]
\&      };
\&    
\&      const manochrome = (MANochrome.has_index_id()) ? 
\&        new MANochrome_pod(options) : false;
\&    
\&    })();
\&  </script>
.Ve
.SH "INSTALLATION"
.IX Header "INSTALLATION"
Just type
.PP
.Vb 1
\&  make
.Ve
.PP
to install and
.PP
.Vb 1
\&  make uninstall
.Ve
.PP
to uninstall.
.SH "FILES"
.IX Header "FILES"
.IP "\fI/usr/local/bin/manopod2html\fR" 4
.IX Item "/usr/local/bin/manopod2html"
.PD 0
.IP "\fI/usr/local/share/man/man1/manochrome.1\fR" 4
.IX Item "/usr/local/share/man/man1/manochrome.1"
.IP "\fI/usr/local/share/man/man1/manopod2html.1\fR" 4
.IX Item "/usr/local/share/man/man1/manopod2html.1"
.IP "\fI/usr/local/share/manochrome/js/manochrome.js\fR" 4
.IX Item "/usr/local/share/manochrome/js/manochrome.js"
.IP "\fI/usr/local/share/manochrome/js/init.js\fR" 4
.IX Item "/usr/local/share/manochrome/js/init.js"
.IP "\fI/usr/local/share/manochrome/css/manochrome_dark.css\fR" 4
.IX Item "/usr/local/share/manochrome/css/manochrome_dark.css"
.IP "\fI/usr/local/share/manochrome/css/manochrome_light.css\fR" 4
.IX Item "/usr/local/share/manochrome/css/manochrome_light.css"
.IP "\fI/usr/local/share/manochrome/css/manochrome_monk.css\fR" 4
.IX Item "/usr/local/share/manochrome/css/manochrome_monk.css"
.PD
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fBman\fR\|(1),
\&\fBpod2html\fR\|(1),
\&\fBrman\fR\|(1),
\&\fBpod2man\fR\|(1),
\&\fBgroff\fR\|(1)
.SH "AUTHORS"
.IX Header "AUTHORS"
Albert Mendes <mailto:tray.mendes@gmail.com>
<https://github.com/aaamendes>
.SH "COPYRIGHT"
.IX Header "COPYRIGHT"
\&\fICopyright\ X\ 2024\ Albert\ Mendes\fR
.PP
MANochrome is free software: you can redistribute it and/or modify
it under the terms of the \s-1GNU\s0 General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.
.PP
This program is distributed in the hope that it will be useful,
but \s-1WITHOUT ANY WARRANTY\s0; without even the implied warranty of
\&\s-1MERCHANTABILITY\s0 or \s-1FITNESS FOR A PARTICULAR PURPOSE.\s0  See the
\&\s-1GNU\s0 General Public License for more details.
.PP
You should have received a copy of the \s-1GNU\s0 General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.


js/manochrome.js
class MANochrome {
  constructor(options={}) {
    this.html = document.body;
    this.generator = (typeof(options.generator) === "string") ? options.generator : "pod2html";
    this.options = options;
    this.headers = null;
    this.index = null;
    this.index_links = null;
    this.navigation = null;
    this.navigation_active_css_class = "nav-link-active";
    this.navi_open = true;
    this.windowMaxWidth = 430;

    this.build_index()
      .then((resp) => this.separate_index_content())
      .then((resp) => this.prettify_navigation())
      .then((resp) => { return this.sort_headers(); })
      .then((resp) => { 
        this.headers = resp;

        this.insert_open_close()
          .then(this.insert_links())
          .then(this.insert_logo())
          .then(this.set_mobile())
          .then(this.set_options());

      });
  }

  set_options() {
    return new Promise((resolve, reject) => {
      resolve(true);
    });
  }

  set_mobile() {
    return new Promise((resolve, reject) => {
      if (window.innerWidth <= this.windowMaxWidth) {
        document.getElementById("open-close-link").click();
      }
      resolve(true);
    });
  }

  navi_open_close(link) {
    let navi = document.getElementById("navigation");
    let content = document.getElementsByClassName("content")[0];
    if (this.navi_open) {
      navi.classList.add("navi-hide");
      content.classList.add("c-move-left");
      link.classList.add("navi-closed-link");
      this.navi_open = !this.navi_open;
    }
    else {
      navi.classList.remove("navi-hide");
      content.classList.remove("c-move-left");
      link.classList.remove("navi-closed-link");
      this.navi_open = !this.navi_open;
    }
  }

  insert_open_close() {
    return new Promise((resolve, reject) => {
      let ocwrapper = document.createElement("div");
      ocwrapper.classList.add("open-close-wrapper");
      ocwrapper.setAttribute("id", "open-close-wrapper");

      let a = document.createElement("a");
      a.setAttribute("id", "open-close-link");
      a.setAttribute("href", "#");

      a.addEventListener("click", (e) => {
        e.preventDefault();
        this.navi_open_close(a);
      });

      ocwrapper.appendChild(a);

      let navi = document.getElementById("navigation");
      navi.insertBefore(ocwrapper, navi.firstChild);
    });
  }

  insert_links() {
    return new Promise((resolve, reject) => {
      if (typeof(this.options.links) !== "object") {
        resolve(false);
        return;
      }
      let wrapper = document.createElement("div");
      wrapper.setAttribute("id", "links-wrapper");

      let h2 = document.createElement("h2");
      h2.textContent = "🔗 LINKS";
      let ul = document.createElement("ul")

      this.options.links.map((item, i) => {
        let li = document.createElement("li");
        let a = document.createElement("a");
        a.innerHTML = item.html;
        a.setAttribute("href", item.url);
        a.setAttribute("target", "_blank");
        li.appendChild(a);
        ul.appendChild(li);
      });

      wrapper.appendChild(h2);
      wrapper.appendChild(ul);

      let navi = document.getElementById("navigation");
      navi.appendChild(wrapper);

      resolve(true);
    });
  }

  insert_logo() {
    return new Promise((resolve, reject) => {
      if (typeof(this.options.logo_path) !== "string") {
        resolve(false);
        return;
      }

      let a = document.createElement("a");
      a.classList.add("manochrome-logo-link");
      a.setAttribute("href", "#");

      let navigation_node = document.getElementById("navigation");
      let img = document.createElement("img");
      img.setAttribute("src", this.options.logo_path);
      img.setAttribute("id", "manochrome-logo");
      img.style.width = "100%";

      a.appendChild(img);

      navigation_node.insertBefore(a, navigation_node.firstChild);

      resolve(true);
    });
  }

  build_index() {
    return new Promise((resolve, reject) => {
      let msg = `method build_index() not implemented in class ${this.constructor.name}`
      console.warn(msg);
      resolve(msg);
    });
  }

  separate_index_content() {
    return new Promise((resolve, reject) => {

      this.body = document.body.cloneNode(true);
      this.wrapper = document.createElement("div");
      this.wrapper.classList.add("wrapper");

      this.content = document.createElement("div");
      this.content.classList.add("content");
      this.content.innerHTML = this.body.innerHTML;

      this.navigation = document.createElement("div");
      this.navigation.classList.add("navigation");
      this.navigation.setAttribute("id", "navigation");
      this.navigation.innerHTML = this.index.outerHTML;

      [
        this.navigation,
        this.content
      ].map((node, index) => {
        this.wrapper.appendChild(node);
      });

      document.body.innerHTML = this.wrapper.outerHTML;

      // Get the cloned index.
      this.index = document.getElementById("index");

      resolve("separate_index_content done");
    });
  }

  get_nav_index_by_string(str, get_num=0) {
    return new Promise((resolve, reject) => {
      for (let i = 0; i < this.index_links.length; i++) {
        if (this.index_links[i].innerHTML === str) {
          if (get_num === 0)
            resolve(this.index_links[i]);
          else
            resolve(i);

          return;
        }
      }
    });

    resolve(false);
    return;
  }

  async change_nav_on_scroll() {
    let offset_y = window.scrollY;
    let grace = 21;

    for (let i = 0; i < this.headers.length; i++) {
      let check_node = await this.get_nav_index_by_string(this.headers[i].innerHTML);

      if (offset_y >= (this.headers[i].offsetTop - grace)) {
        this.set_link_active(check_node);
      }

    }
  }

  set_link_active(node) {
    for (let i = 0; i < this.index_links.length; i++) {
      this.index_links[i].classList.remove(this.navigation_active_css_class);
    }

    node.classList.add(this.navigation_active_css_class);
  }

  async sort_headers() {
    let new_arr = [];
    for (let i = 0; i < this.headers.length; i++) {
      let link_index = await this.get_nav_index_by_string(this.headers[i].innerHTML, 1);
      new_arr[link_index] = this.headers[i];
    }

    return new_arr;
  }

  prettify_navigation() {
    return new Promise((resolve, reject) => {
      this.index_links = this.index.getElementsByTagName("a");

      for (let i = 0; i < this.index_links.length; i++) {
        this.index_links[i].addEventListener("click", (e) => {
          this.set_link_active(this.index_links[i]);
        });
      }

      this.set_link_active(this.index_links[0]);

      this.headers = Array.from(document.getElementsByTagName("h1"))
      this.headers = this.headers.concat(Array.from(document.getElementsByTagName("h2")));
      this.headers = this.headers.concat(Array.from(document.getElementsByTagName("h3")));
      this.headers = this.headers.concat(Array.from(document.getElementsByTagName("h4")));

      window.addEventListener("scroll", (e) => {
        this.change_nav_on_scroll();
      });

      resolve("prettify_navigation done");
    });
  }

  /*
   * HTML generated by pod2html has an UL with id "index".
   */
  static has_index_id() {
    return (document.getElementById("index") !== null) ? (
      (document.getElementById("index").tagName === "UL") ? true : false
    ) : false;
  }
  /*
   * So far in development, we only have pod2html.
   */
  static determine_generator() {
    const gen_try = document.head.querySelector(
      'meta[name="generator"]'
    );

    return (gen_try === null) ? "pod2html" :
      (
        (gen_try.content.includes("groff")) ? "groff" : gen_try.content
      )
  }
}

class MANochrome_pod extends MANochrome {
  constructor(options={}) {
    super(options);
  }

  /*
   * Index is already there.
   * Clone Node (<UL>) and remove original.
   */
  build_index() {
    return new Promise((resolve, reject) => {
      const index = document.getElementById("index");
      this.index = index.cloneNode(true);
      index.remove();
      resolve("build_index done");
    });
  }
}

/*
 * TODO
 * This is experimental.
 */
class MANochrome_groff extends MANochrome {
  constructor(generator) {
    super(generator);
  }

  build_index() {
    let ul = document.createElement("ul");
    ul.setAttribute("id", "index");
    let delete_arr = [];

    return new Promise((resolve, reject) => {
      let next_link = document.getElementsByTagName("h1")[0];
      let i = 0;
      while (next_link.nextElementSibling.tagName !== "HR") {
        if (next_link.nextElementSibling.tagName !== "A") {
          next_link = next_link.nextElementSibling;
          delete_arr.push(next_link);
          continue;
        }

        let li = document.createElement("li");
        let a = next_link.nextElementSibling.cloneNode(true);
        li.appendChild(a);
        ul.appendChild(li);

        next_link = next_link.nextElementSibling;
        delete_arr.push(next_link);
        i++;
      }

      /*
       * Set this.index.
       */
      this.index = ul;

      delete_arr.map((node, i) => {
        node.remove();
      });

      /* 
       * Remove HR.
       */
      const hr = document.getElementsByTagName("h1")[0].nextElementSibling;

      if (hr.tagName === "HR")
        hr.remove();

      resolve("build_index done");
    });
  }
}
